File: typescript/Text.ts
declare function createSwiftText(id: string, text: string, style: TextStyle): void;
declare function updateSwiftText(id: string, text: string, style: TextStyle): void;

export interface TextStyle {
    fontSize?: number;
    fontWeight?: string;
    color?: string;
    backgroundColor?: string;
    padding?: number;
    cornerRadius?: number;
    italic?: boolean;
    underline?: boolean;
    strikethrough?: boolean;
    kerning?: number;
    lineSpacing?: number;
    textAlignment?: string;
    shadowRadius?: number;
    shadowX?: number;
    shadowY?: number;
    shadowColor?: string;
}

export interface TextConfig {
    id?: string;
    text: string;
    style?: TextStyle;
}

const defaultStyle: TextStyle = {
    fontSize: 32,
    fontWeight: 'regular',
    color: '#000000',
    backgroundColor: '#ffffff',
    padding: 8,
    cornerRadius: 8,
    italic: false,
    underline: false,
    strikethrough: false,
    kerning: 0,
    lineSpacing: 0,
    textAlignment: 'leading'
};

// Text styling API
export class Text {
    private textIds: Set<string>;
    private id: string;
    private text: string;
    private style: TextStyle;

    constructor(config: TextConfig, textIds: Set<string>) {
        this.textIds = textIds;
        this.id = config.id || generateId();
        this.text = config.text;
        this.style = {
            ...defaultStyle,
            ...config.style || {}
        };
    }

    // Create the text in SwiftUI
    create(): this {
        createSwiftText(this.id, this.text, this.style);
        this.textIds.add(this.id);
        return this;
    }

    // Update existing text
    update(config: Partial<TextConfig>): this {
        if (config.text !== undefined) {
            this.text = config.text;
        }
        if (config.style) {
            this.style = {
                ...this.style,
                ...config.style
            };
        }
        updateSwiftText(this.id, this.text, this.style);
        return this;
    }

    // Get the current ID
    getId(): string {
        return this.id;
    }
}

// Helper function to generate unique IDs
function generateId(): string {
    return 'text_' + Math.random().toString(36).substr(2, 9);
}

File: JS/Text.swift
//
//  Text.swift
//  JS
//
//  Created by Tyler Williams on 3/2/25.
//

import SwiftUI

struct TextStyle: Equatable {
    var fontSize: CGFloat = 16
    var fontWeight: Font.Weight = .regular
    var foregroundColor: Color = .primary
    var backgroundColor: Color = .clear
    var padding: CGFloat = 8
    var cornerRadius: CGFloat = 8
    var italic: Bool = false
    var underline: Bool = false
    var strikethrough: Bool = false
    var kerning: CGFloat = 0
    var lineSpacing: CGFloat = 0
    var textAlignment: TextAlignment = .leading
    var shadow: (radius: CGFloat, x: CGFloat, y: CGFloat, color: Color)? = nil
    
    static func == (lhs: TextStyle, rhs: TextStyle) -> Bool {
        lhs.fontSize == rhs.fontSize &&
        lhs.fontWeight == rhs.fontWeight &&
        lhs.padding == rhs.padding &&
        lhs.cornerRadius == rhs.cornerRadius &&
        lhs.italic == rhs.italic &&
        lhs.underline == rhs.underline &&
        lhs.strikethrough == rhs.strikethrough &&
        lhs.kerning == rhs.kerning &&
        lhs.lineSpacing == rhs.lineSpacing &&
        lhs.textAlignment == rhs.textAlignment &&
        lhs.shadow?.radius == rhs.shadow?.radius &&
        lhs.shadow?.x == rhs.shadow?.x &&
        lhs.shadow?.y == rhs.shadow?.y
    }
}

struct TextItem: Identifiable {
    let id: String
    var text: String
    var style: TextStyle
}
